### PR TITLE
[WIP] Add Data Fixtures for Lambda Event Types

### PR DESCRIPTION
**Pull request by naftulikay**
Tests and documentation forthcoming. I am using these data structures for my serverless event bus in Rust using Crowbar.

My lambda handler looks something like this:

```rust
lambda!(|event, _context| {
    // dispatch event based on its type
    match serde_json::from_value::<Event>(event) {
        Ok(Event::Auth(event)) => Ok(to_value(&service::auth::route(&context, &event)).unwrap()),
        Ok(Event::CloudWatch(event)) => Ok(service::cloudwatch::route(&context, &event)),
        Ok(Event::Http(event)) => Ok(to_value(&service::http::route(&context, &event)).unwrap()),
        Ok(Event::Records(records)) => Ok(service::multi::route_all(&context, records.entries)),
        Ok(Event::Unknown(event)) => Ok(service::unknown::route(&context, &event)),
        Err(_) => {
            error!("Unable to convert event.");
            Ok(json!({ "statusCode": 400 ,"message": "Unable to convert event." }))
        }
    }
});
```

Thus, I do routing based on event type. For example, SNS:

```rust
pub fn route(context: &Context, record: &sns::Record) {
    info!("Received SNS Event: {}", record.event.message_id);

    match from_str(&record.event.message) {
        Ok(SnsEventKind::Ses(message)) => service::ses::receive(&context, &message),
        Ok(SnsEventKind::Unknown(value)) => warn!("Unknown JSON Event: {}", to_string_pretty(&value).unwrap()),
        Err(_) => error!("Unknown non-JSON event: {}", record.event.message),
    };
}
```

My multi-event router:

```rust
/// Dispatch multiple events.
pub fn route_all(context: &Context, events: Vec<Record>) -> Value {
    for event in events {
        match event {
            Record::S3(event_record) => s3::route(&context, &event_record),
            Record::Ses(event_record) => {
                // if it's an SES request/response, a response may be required, so only process one
                // of them if the type is RequestResponse
                match event_record.event.receipt.action {
                    Action::Lambda(ref action) if action.invocation_type == RequestResponse => {
                        return to_value(ses::filter(&context, &event_record)).unwrap();
                    },
                    _ => ses::route(&context, &event_record.event),
                };
            },
            Record::Sns(event_record) => sns::route(&context, &event_record),
            Record::Unknown(value) => {
                error!("Unknown event record type: {}", to_string_pretty(&value).unwrap())
            }
        };
    }

    Value::Null
}
```

This dispatching occurs in around 400-500 microseconds and works extremely well to make things type safe.

Obviously there's a lot here and it's not done. If @ilianaw or others have any input, RFC is open to comments :smiley:

### CODE REVIEW

**Comment by ilianaw** [comment]
I like this. :)

Can you add what has been done and what remains [as checkboxes](https://help.github.com/articles/about-task-lists/) to the pull request description so we can see the path to this leaving a WIP state?

**Comment by naftulikay** [comment]
:tada: so happy that you're welcoming this PR. I will create a separate issue and track things there. There are _so_ many different data types to support, so this PR will maybe only cover a subset with future on the way, hope that's okay.

**Comment by ilianaw** [comment]
Yeah, I don't have an issue with incremental development over time.

I'm hoping I'll have a chance to do a good read through of all of this soon.

**Comment by naftulikay** [comment]
@ilianaw can you please go turn on Travis CI for this project? I have updated with a build badge and Travis configuration for building and testing.

**Comment by naftulikay** [comment]
Latest adds support for CloudWatch AutoScaling events.

**Comment by softprops** [comment]
I'd love to see this branch ship fast with additional followups to add other events types. Commented with extra thoughts on the checkbox ticket

File: src/data/apigateway/mod.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            HttpMethod::OPTIONS => "OPTIONS",
+            HttpMethod::DELETE => "DELETE"
+        })
+    }
+}
+
+#[derive(Serialize,Deserialize)]
+pub struct HttpEvent {
+    pub body: Option<String>,
+    pub headers: Option<HashMap<String, String>>,
```
Comment (softprops) [inline src/data/apigateway/mod.rs]: One small thought. I did something similar initially in a separate project and found that the extra option type to be a bit annoying from a usability perspective. Espectially those new to rust

```
9 |     println!("{:?}", ev.query_string_parameters.and_then(|m| m.get("foo")));
  |     ---------------------------------------------------------^--------------
  |     |                                                        |          |
  |     |                                                        |          `m` dropped here while still borrowed
  |     |                                                        borrowed value does not live long enough
  |     borrowed value needs to live until here
```

Was't a very great experience.

I found a way to [deserialize the nullable maps into default values](https://github.com/softprops/lando/blob/master/src/request.rs#L15-L16) greatly improved usability for new comers

as well as reduced verbosity while retaining the same level of rust runtime saftey.

```rust
 println!("{:?}", ev.query_string_parameters.and_then(|m| m.get("foo")))
```

```rust
println!("{:?}", ev.query_string_parameters.get("foo"));
```

File: src/data/apigateway/mod.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            body = match self.body {
+                Some(_) => true,
+                None => false
+            }
+        )
+    }
+}
+
+#[derive(Serialize,Deserialize)]
+pub struct HttpEventRequestContext {
```
Comment (softprops) [inline src/data/apigateway/mod.rs]: I think you can largely get a way with a struct level `#[serde(rename_all = "camelCase")]` here

File: src/data/apigateway/mod.rs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    pub status: HttpStatus,
+    pub headers: HashMap<String, String>,
+    pub body: Option<String>,
+    #[serde(rename="isBase64Encoded",default)]
+    pub is_base64: bool,
+}
+
+impl HttpResponse {
+
+    pub fn empty(status: HttpStatus) -> Self {
```
Comment (softprops) [inline src/data/apigateway/mod.rs]: I'm trying to convince myself to drop an existing external effort to add types to crowbar. Something I'd be missing here is a [Default impl](https://github.com/softprops/lando/blob/57cf8b4d5884c0a5e5035a316dadfa1dc596c160/src/response.rs#L29-L33)

File: src/data/apigateway/mod.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+impl HttpEventRequestContext {
+
+    pub fn time(&self) -> Option<DateTime<Utc>> {
+        // Utc::datetime_from_str(&self.request_time, "").ok()e
+        None
+    }
+}
+
+serde_aux_enum_number_declare!(HttpStatus {
```
Comment (softprops) [inline src/data/apigateway/mod.rs]: I'd maybe suggest not going too overboard with type mappings here. I see crowbar being something as a low level mapping to raw primitive types, in this case a u16 would do. The reasoning behind this line of thinking is that you could then more easily adapt to and not duplicate efforts to create a typed interface like [http](https://github.com/hyperium/http/blob/master/src/status.rs#L305-L497) on top of crowbar. As is you couldn't build a mapping to the http crate on top of this because if the incomplete mapping of http statuses. I'd much prefer the crowbar retain the most primitive types and enable layering of extra layers on top in another crate.

File: src/data/apigateway/mod.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    pub resource_id: String,
+    #[serde(rename="resourcePath")]
+    pub resource_path: String,
+    pub stage: String,
+}
+
+impl HttpEventRequestContext {
+
+    pub fn time(&self) -> Option<DateTime<Utc>> {
+        // Utc::datetime_from_str(&self.request_time, "").ok()e
```
Comment (softprops) [inline src/data/apigateway/mod.rs]: you could probably ditch a hard coupling of crowbar to a [crate](https://github.com/ilianaw/rust-crowbar/pull/31/files#diff-80398c5faae3c069e4e6aa2ed11b28c0R17) by letting it be a user decision to select a time parsing crate and let crowbar be smaller.

File: src/data/apigateway/mod.rs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    pub resource: String,
+    #[serde(rename="requestContext")]
+    pub request_context: HttpEventRequestContext,
+    #[serde(rename="stageVariables")]
+    pub stage_variables: Option<HashMap<String, String>>,
+}
+
+impl HttpEvent {
+
+    pub fn get_header(&self, key: &str) -> Option<&str> {
```
Comment (softprops) [inline src/data/apigateway/mod.rs]: There are some some growing community recommendations on avoiding get_ where you can https://rust-lang-nursery.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Instead a simple `.header("Foo")` will do

File: src/data/ses/mod.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    Continue,
+    #[serde(rename="STOP_RULE")]
+    StopRule,
+    #[serde(rename="STOP_RULE_SET")]
+    StopRuleSet,
+}
+
+impl Response {
+
+    pub fn from(action: ResponseAction) -> Self {
```
Comment (softprops) [inline src/data/ses/mod.rs]: Why not impl the From trait instead?

File: src/data/apigateway/mod.rs (hunk 6 above)
Comment (naftulikay) [reply src/data/apigateway/mod.rs]: Roger that.

File: src/data/ses/mod.rs (hunk 7 above)
Comment (naftulikay) [reply src/data/ses/mod.rs]: Let's do it :+1:

File: src/data/apigateway/mod.rs (hunk 5 above)
Comment (naftulikay) [reply src/data/apigateway/mod.rs]: How can we genericize this to work like that?

File: src/data/apigateway/mod.rs (hunk 4 above)
Comment (naftulikay) [reply src/data/apigateway/mod.rs]: So should we rely on another crate for this? I do actually like having strongly typed things.

File: src/data/apigateway/mod.rs (hunk 3 above)
Comment (naftulikay) [reply src/data/apigateway/mod.rs]: :+1:

File: src/data/apigateway/mod.rs (hunk 1 above)
Comment (naftulikay) [reply src/data/apigateway/mod.rs]: Yes, let's deserialize things like maps and vectors into default (empty) values.

**Comment by naftulikay** [comment]
I will start breaking this apart into a PR per service supported, I think that's probably the most prudent course of action.

One thing that is probably a really good idea is to find a way to efficiently nest CloudWatch style events and deserialize them more efficiently than I'm currently doing.

File: src/data/apigateway/mod.rs (hunk 4 above)
Comment (softprops) [reply src/data/apigateway/mod.rs]: I'd actually not recommend adding any more crate dependencies on crowbar. I don't see crowbar as a framework but as an faithful representation of lambda. I'm thinking now it its really api gateway, the service, that defines what status codes you can dispatch lambdas on so that may actually be a more restrictive set. I redact my previous comment!

File: src/data/apigateway/mod.rs (hunk 5 above)
Comment (softprops) [reply src/data/apigateway/mod.rs]: If we didn't want to represent this as a string that lambda events do, you could maybe use a feature + cfg. I don't feel really strongly about this. Only that I'd like to see crowbar have a solve only the minimal set of problems with a minimal set of dependencies. Motivations are that it lets users pick a crate they prefer for representing time typically so that they only have one crate for dealing with time in their projects ( think about what hyper did with tls ) and it frees up crowbar up from a bit of maintenance by not worrying falling behind on its dependencies. I also don't feel too strongly here but that's my two cents. I'd feel otherwise if rust had a std lib way of representing timestamps.

**Comment by softprops** [comment]
> One thing that is probably a really good idea is to find a way to efficiently nest CloudWatch style events and deserialize them more efficiently than I'm currently doing.

Can you link to an example in this pull or sketch out in a comment what you mean?

File: src/data/apigateway/mod.rs (hunk 2 above)
Comment (softprops) [reply src/data/apigateway/mod.rs]: note: comments like this are not really blockers for me and could easily fit into followups

File: src/data/apigateway/mod.rs (hunk 1 above)
Comment (softprops) [reply src/data/apigateway/mod.rs]: I really think that kind of thing helps the ergonomics for people using this.

File: src/data/apigateway/mod.rs (hunk 4 above)
Comment (ilianaw) [reply src/data/apigateway/mod.rs]: I agree that a u16 would do for the HTTP status.

I experimented a lot with the `http` crate for another project involving Lambda@Edge. It helped with initial development but I eventually ripped it out because it was doing a lot of validation that CloudFront already does for you.

File: src/data/apigateway/mod.rs (hunk 5 above)
Comment (ilianaw) [reply src/data/apigateway/mod.rs]: I don't think genericizing the time crate is necessary to land this PR.

**Comment by ilianaw** [comment]
Overall I'm happy with this, and thanks to @softprops for the additional review.

I'm fine with continuing with this PR, or breaking it apart as a PR per service (the latter would be great as its easier for others to help write the structs out).

I have two thoughts:

* What happens when AWS inevitably adds another field to these objects? Does serde correctly ignore extra fields?
* Would it be reasonable to make all of these an enabled-by-default feature to allow folks who aren't using this to disable it and speed up compile times?

**Comment by softprops** [comment]
> What happens when AWS inevitably adds another field to these objects? Does serde correctly ignore extra fields?

Yes. This used to not be true but is now a default of serde.

**Comment by naftulikay** [comment]
I apologize for my lateness, just getting back from vacation. I will get caught up this week.

**Comment by ilianaw** [comment]
No worries! ^_^;

**Comment by naftulikay** [comment]
TL;DR the events received from Lambda can come via a few sources:

 - SNS: a fairly consistent data type; you'd basically just deserialize the `message` property into the struct(s) that you expect.
 - CloudWatch: there are a lot (#32) of different CloudWatch data types.
 - API Gateway: requests, auth requests, and responses.

Therefore, our enums for representing all kinds of incoming events should basically include these three, and then users can disambiguate between the implementations (ie `cloudwatch::BatchEvent`, `cloudwatch::ApiCallEvent`, `apigateway::Request`, `apigateway::AuthRequest`, etc.)

I will probably break this PR down into tiny pieces and make one PR for each service to add. I wholeheartedly agree with removing `Option<T>` wherever possible and using the default for things like `Vec<T>` and `BTreeMap<K, V>`.

**Comment by softprops** [comment]
Is there any way to help move this forward? I'd love to help
